src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,231 @@
+     */
+    case ScanType::OpEquals:
+      if (lower_bound == upper_bound) {
+        return position_list;
+      }
+      break;
+    case ScanType::OpNotEquals:
+      if (lower_bound == upper_bound)
+        scan_op = [](auto l, auto r) { return true; };
+      break;
+    case ScanType::OpLessThanEquals:
+      scan_op = [&](auto l, auto r) { return l < upper_bound; };
+      break;
+    case ScanType::OpGreaterThan:
+      scan_op = [&](auto l, auto r) { return l >= upper_bound; };
+      break;
+      // Rest stays as it is.
+    default:
+      break;
+  }
+
+  const auto attr_vector = segment->attribute_vector();
+  for (auto index = ChunkOffset{0}; index < segment->size(); ++index) {
+    const auto value_id = attr_vector->get(index);
+    if (value_id != segment->null_value_id() && scan_op(value_id, lower_bound)) {
+      position_list->push_back(RowID{chunk_id, index});
+    }
+  }
+
+  return position_list;
+}
+
+template <typename T>
+std::shared_ptr<PosList> TableScan::_tablescan_value_segment(std::shared_ptr<ValueSegment<T>> segment,
+                                                             ChunkID chunk_id) {
+  const auto scan_op = _create_scan_operation<T>();
+  const auto values = segment->values();
+
+  auto position_list = std::make_shared<PosList>();
+  const auto search_val = type_cast<T>(_search_value);
+
+  auto index = ChunkOffset{0};
+  for (const auto& value : values) {
+    if (!segment->is_null(index) && scan_op(value, search_val)) {
+      position_list->push_back(RowID{chunk_id, index});
+    }
+    ++index;
+  }
+
+  return position_list;
+}
+
+template <typename T>
+std::shared_ptr<PosList> TableScan::_tablescan_reference_segment(std::shared_ptr<ReferenceSegment> segment,
+                                                                 ChunkID chunk_id) {
+  auto position_list = std::make_shared<PosList>();
+
+  const auto input_position_list = segment->pos_list();
+  const auto table = segment->referenced_table();
+  const auto search_val = type_cast<T>(search_value());
+  const auto scan_op = _create_scan_operation<T>();
+
+  for (auto index = ChunkOffset{0}; index < segment->size(); ++index) {
+    const auto row = (*input_position_list)[index];
+    const auto chunk = table->get_chunk(row.chunk_id);
+    const auto target_segment = chunk->get_segment(_column_id);
+
+    const auto dict_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(target_segment);
+    const auto val_segment = std::dynamic_pointer_cast<ValueSegment<T>>(target_segment);
+
+    Assert(dict_segment || val_segment, "Segment that ReferenceSegement references is not supported by TableScan.");
+
+    if (val_segment) {
+      // If value is null, it cannot appear in result set, so just continue.
+      if (val_segment->is_null(row.chunk_offset)) {
+        continue;
+      }
+      const auto values = val_segment->values();
+      const auto casted_value = type_cast<T>(values[row.chunk_offset]);
```

Comment:
Shouldn't the values of the ValueSegment already have `T` as type? Why are you type casting again?